internal/storagenode/logstream/executor.go

```diff
@@ -615,28 +619,42 @@ func (lse *Executor) isPrimary() bool {
 	return len(lse.primaryBackups) > 0 && lse.primaryBackups[0].StorageNodeID == lse.snid && lse.primaryBackups[0].LogStreamID == lse.lsid
 }
 
-func (lse *Executor) restoreLogStreamContext(recoveryPoints storage.RecoveryPoints) *logStreamContext {
+func (lse *Executor) restoreLogStreamContext(rp storage.RecoveryPoints) *logStreamContext {
+	cc := rp.LastCommitContext
+	first := rp.CommittedLogEntry.First
+	last := rp.CommittedLogEntry.Last
+
 	lsc := newLogStreamContext()
-	commitVersion, globalHighWatermark, uncommittedLLSNBegin := lsc.reportCommitBase()
-	uncommittedLLSNEnd := lsc.uncommittedLLSNEnd.Load()
-	localLowWatermark := lsc.localLowWatermark()
-	localHighWatermark := lsc.localHighWatermark()
-
-	if lastCC := recoveryPoints.LastCommitContext; lastCC != nil {
-		commitVersion = lastCC.Version
-		globalHighWatermark = lastCC.HighWatermark
-	}
-	if boundaries := recoveryPoints.CommittedLogEntry; boundaries.First != nil {
-		lastLLSN := boundaries.Last.LLSN
-		uncommittedLLSNBegin = lastLLSN + 1
-		uncommittedLLSNEnd = lastLLSN + 1
-		localLowWatermark = *boundaries.First
-		localHighWatermark = *boundaries.Last
-	}
-
-	lsc.storeReportCommitBase(commitVersion, globalHighWatermark, uncommittedLLSNBegin)
-	lsc.uncommittedLLSNEnd.Store(uncommittedLLSNEnd)
-	lsc.setLocalLowWatermark(localLowWatermark)
-	lsc.setLocalHighWatermark(localHighWatermark)
+	restoreMode := "init"
+	defer func() {
+		lse.logger.Info("restore log stream context", zap.String("mode", restoreMode))
+	}()
+
+	if cc == nil && last == nil { // new log stream replica
+		return lsc
+	}
+
+	if cc != nil && last != nil { // recovery
+		restoreMode = "recovered"
+		uncommittedOffset := cc.CommittedLLSNBegin + types.LLSN(cc.CommittedGLSNEnd-cc.CommittedGLSNBegin)
+		if uncommittedOffset-1 == last.LLSN {
+			lsc.storeReportCommitBase(cc.Version, cc.HighWatermark, uncommittedOffset)
+			lsc.uncommittedLLSNBegin.Store(uncommittedOffset)
+			lsc.uncommittedLLSNEnd.Store(uncommittedOffset)
+			lsc.setLocalLowWatermark(*first)
+			lsc.setLocalHighWatermark(*last)
+			return lsc
+		}
+	}
+
+	// something wrong
+	restoreMode = "invalid"
```

Comment:
Is it correct to mean that stored `cc` and `commit` are inconsistent?
Is it for L652 not to report and for L654 - L657 to inform the recovery starting point during `SyncInit'?

Reply:
Yes, they are. 

There may be the following situations:

- Log entries exist, but no commit context.
- The last log entry's sequence number is not equal to one computed by commit context.

They may happen due to restarting during the synchronization of a replica.